src/client.ts

```diff
@@ -43,39 +32,93 @@ export type AcmeGroup = Opaque<any, 'AcmeGroup'>;
-          if (res.statusCode !== 200) {
-            reject(new Error('Provider authentication failed'));
-          } else {
-            resolve();
-          }
-        },
-      );
+  private withBaseUri(path: string): string {
+    return `https://${this.config.cybereasonHost}:${this.config.cybereasonPort}/${path}`;
+  }
+
+  private async getRequest(uri: string, cookie: string): Promise<Response> {
+    const response = await fetch(uri, {
+      headers: {
+        cookie,
+      },
     });
+    if (!response.ok) {
+      throw new IntegrationProviderAPIError({
+        endpoint: uri,
+        status: response.status,
+        statusText: response.statusText,
+      });
+    }
+    return response;
+  }
 
-    try {
-      await request;
-    } catch (err) {
-      throw new IntegrationProviderAuthenticationError({
-        cause: err,
-        endpoint: 'https://localhost/api/v1/some/endpoint?limit=1',
-        status: err.status,
-        statusText: err.statusText,
+  private async postRequest(
+    uri: string,
+    body: any,
+    dontRedirect?: boolean,
+  ): Promise<Response> {
+    const redirect = dontRedirect ? 'manual' : 'follow';
+
+    const response = await fetch(uri, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: querystring.stringify(body),
+      redirect,
+    });
+    if (!(response.ok || (response.status === 302 && dontRedirect))) {
+      throw new IntegrationProviderAPIError({
+        endpoint: uri,
+        status: response.status,
+        statusText: response.statusText,
       });
     }
+    return response;
+  }
+
+  public async login(): Promise<string> {
+    const loginRoute = this.withBaseUri('login.html');
+    const res = await this.postRequest(
+      loginRoute,
+      {
+        username: this.config.cybereasonId,
+        password: this.config.cybereasonPassword,
+      },
+      true,
+    );
+    return (
+      res.headers
+        .get('set-cookie')
+        ?.split('; ')
+        .filter((item) => item.search('JSESSIONID') >= 0)[0] || ''
+    );
+  }
+
+  public async logout(cookie: string): Promise<void> {
+    const logoutRoute = this.withBaseUri('logout');
+    await this.getRequest(logoutRoute, cookie);
+  }
+
+  public async verifyAuthentication(): Promise<void> {
+    const cookie = await this.login();
+
+    // NOTE: This only works with 'System admin' role
+    // no API endpoint is available for all roles, need to find
+    // a better endpoint to test the cookie that is not hindered
+    // by role permissions.
+    const groupsUri = this.withBaseUri('rest/groups');
+    const res = await this.getRequest(groupsUri, cookie);
+
+    // if cookie is invalid, API responds with login.html
+    if (res.headers.get('content-type') === 'text/html')
+      throw new IntegrationProviderAuthenticationError({
+        endpoint: groupsUri,
+        status: res.status,
+        statusText: res.statusText,
+      });
```

Comment:
Problem: API doesn't respond with a 401 error when the cookie is invalid. Hacked a way to check if the response content is HTML instead which shouldn't happen when the cookie is valid.

See supported error codes here:
https://nest.cybereason.com/documentation/api-documentation/all-versions/responses#error-codes

Reply:
Noted! It does feel hacky but sometimes there are edge cases such as this one. Let's keep this one for now and later if we can think of some neater way of handling this, we'll change it :+1: